de.metas.contracts/src/main/java/de/metas/contracts/flatrate/process/C_Flatrate_Term_Change.java

```diff
@@ -69,60 +72,61 @@ protected String doIt()
 		}
 
-		
-		final I_C_Flatrate_Term currentTerm = InterfaceWrapperHelper.create(getCtx(), getRecord_ID(), I_C_Flatrate_Term.class, get_TrxName());
-
 		final ContractChangeParameters contractChangeParameters = ContractChangeParameters.builder()
-																.changeDate(changeDate)
-																.isCloseInvoiceCandidate(true)
-																.terminationMemo(terminationMemo)
-																.terminationReason(terminationReason)
-																.build();
-		
-		Services.get(IContractChangeBL.class).cancelContract(currentTerm, contractChangeParameters);
+				.changeDate(changeDate)
+				.isCloseInvoiceCandidate(true)
+				.terminationMemo(terminationMemo)
+				.terminationReason(terminationReason)
+				.build();
+
+		final Iterable<I_C_Flatrate_Term> flatrateTerms = retrieveSelection(getAD_PInstance_ID());
+		flatrateTerms.forEach(currentTerm -> contractChangeBL.cancelContract(currentTerm, contractChangeParameters));
 		
-		getResult().setRecordToRefreshAfterExecution(TableRecordReference.of(currentTerm));
-
 		return "@Success@";
 	}
 
-	
-
-	@Override
-	protected void prepare()
+	private IQueryBuilder<I_C_Flatrate_Term> createQueryBuilder()
 	{
-		final ProcessInfoParameter[] para = getParametersAsArray();
-
-		for (int i = 0; i < para.length; i++)
+		final IQueryFilter<I_C_Flatrate_Term> userSelectionFilter;
+		if (Ini.isClient())
 		{
-			final String name = para[i].getParameterName();
-			if (para[i].getParameter() == null)
-				;
-			else if (name.equals(I_C_Flatrate_Transition.COLUMNNAME_C_Flatrate_Conditions_Next_ID))
-			{
-				// currently not supported by business logic
-				newSubscriptionId = ((BigDecimal)para[i].getParameter()).intValue();
-			}
-			else if (name.equals(PARAM_ACTION))
-			{
-				action = (String)para[i].getParameter();
-			}
-			else if (name.equals(PARAM_CHANGE_DATE))
-			{
-				changeDate = (Timestamp)para[i].getParameter();
-			}
-			else if (name.equals(PARAM_TERMINATION_REASON))
-			{
-				terminationReason = (String)para[i].getParameter();
-			}
-			else if (name.equals(PARAM_TERMINATION_MEMO))
+			// In case of Swing, preserve the old functionality, i.e. if no where clause then select all
+			userSelectionFilter = getProcessInfo().getQueryFilter();
+		}
+		else
+		{
+			userSelectionFilter = getProcessInfo().getQueryFilterOrElse(null);
+			if (userSelectionFilter == null)
 			{
-				terminationMemo = (String)para[i].getParameter();
+				throw new AdempiereException("@NoSelection@");
 			}
-			else
+		}
+
+		return queryBL
+				.createQueryBuilder(I_C_Flatrate_Term.class, getCtx(), ITrx.TRXNAME_None)
+				.filter(userSelectionFilter)
+				.addOnlyActiveRecordsFilter()
+				.addOnlyContextClient();
+	}
+
+	private final Iterable<I_C_Flatrate_Term> retrieveSelection(final int adPInstanceId)
+	{
+		return new Iterable<I_C_Flatrate_Term>()
+		{
+			@Override
+			public Iterator<I_C_Flatrate_Term> iterator()
 			{
-				log.error("Unknown Parameter: " + name);
+				return Services.get(IQueryBL.class)
```

Comment:
favor using the createQueryBuilder(class), i.e. without ctx and trxName in case of running in current trx

Reply:
more, if possible, make this method static... like all methods with annonymous inner classes or lambdas...